### ISSUE TITLE
Order of coordinates of a multi-part feature that maps to the reverse strand maybe 'wrong' in GenBank file generated from SnapGene file?

### ISSUE THREAD

**Issue by nicolazilio** [biopython/biopython#3655]
### Setup

I am reporting a problem with Biopython version, Python version, and operating
system as follows:

```
3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) 
[GCC 9.3.0]
CPython
Linux-5.4.72-microsoft-standard-WSL2-x86_64-with-debian-bullseye-sid
1.79
```

### Expected behaviour
Cloning programs/webapps such as SnapGene or Benchling should display all features included in a GenBank file that has been converted from a SnapGene file with BioPython.

### Actual behaviour
Cloning programs/webapps such as SnapGene or Benchling do not display all features included in a GenBank file that has been converted from a SnapGene file with BioPython. The problematic features are those that contain multiple elements on the reverse strand, for example `complement(join(3873..3941,3081..3872))`. If I reverse the order of those elements, that is, `complement(join(3081..3872,3873..3941))`, then everything's good, and the feature is shown in my fav cloning programs/webapp. I guess the elements of a feature that maps to the reverse/complement strand should nonetheless be listed in an 'increasing order'? 

Although it's most likely a coincidence, I also noticed that for the feature mentioned above, BioPython fails to include one of the notes that SnapGene instead includes in the exported file (the one at the very bottom of the left panel)

![Screenshot 2021-07-05 170536](https://user-images.githubusercontent.com/2800360/124495090-f428cb80-ddb7-11eb-957e-f8becf652083.png)

### Steps to reproduce

I am attaching the original SnapGene map and those converted to GenBank with SnapGene and BioPython to this message.

[test_plasmid.zip](https://github.com/biopython/biopython/files/6765312/test_plasmid.zip)

**Comment by peterjc** [comment]
Thank you for sharing a test case.

This sounds like it could be two bugs in the SnapGene code, (a) not handling the order of complement strand joins correctly, and (b) not handling multiple note entries.

@gouttegd are you able to take a look please?

**Comment by gouttegd** [comment]
@peterjc I will. You may assign the issue to me.

**Comment by gouttegd** [comment]
I confirm the issue with the ordering of complement strand joins, I'll have a patch in the next few days.

The missing `note` qualifier however is a non-issue; our parser correctly handles multiple qualifier values (though I’ll add a test case to cover that, just to be safe). That “missing” `note` is *not* present in the features section of the source SnapGene file, it is dynamically generated by SnapGene when exporting to GenBank.

Basically instead of generating a composite location (`complement(join(...))`), SnapGene generates a simple location covering the entire span of the feature, then adds this extra `note` qualifier to represent the segments that make up the feature.

Why do they do that is beyond me, but I do not plan to emulate this behaviour. I’ll fix the parser so that we end up with properly ordered joins, so that all the information will be in the feature location (where it should be) without the need for an extra note.

**Comment by peterjc** [comment]
Sounds good to me, thank you.

**Comment by nicolazilio** [comment]
Great, thanks! @gouttegd, the only advantage that I see with having that additional note is that they include the name/label of each part of the main parent feature. In the case above, for example, `signal sequence`. Would it be possible to preserve that information in some way in some other way, as it could be useful?

**Comment by gouttegd** [comment]
As far as I know there is no provision in the `SeqRecord` model (or in the GenBank model for that matter) to represent "named sub features", so any solution we could come up with to keep that information would basically be a kludge of some sort.

Possible solution A: The "SnapGene method", where we generate a note to record everything we cannot record otherwise:

```
     CDS             complement(join(3081..3872,3873..3941))
                     /product="beta-lactamase"
                     /note="part 3873-3941: signal sequence"
```

That's definitely a kludge, and I am *very* reluctant to do that.

Possible solution B: Generate a distinct feature with its own `label`:

```
     CDS             complement(join(3081..3872,3873..3941))
                     /product="beta-lactamase"
                     [...]
     misc_feature    complement(3873..3941)
                     /label="signal sequence"
```

That's cleaner and actually less of a kludge. Would that solution be satisfying?

But really, to me it seems that SnapGene is abusing the concept of compound features by allowing each part to have a name/label. If a part of a compound feature is distinct enough to have its own label, then I believe it should really be a distinct feature, not a "sub feature".

**Comment by nicolazilio** [comment]
I think the reason why SnapGene does it that way may be related to the aesthetics of displaying such compound features, see an example below not related to the feature I used nefore

![Untitled-1 copy](https://user-images.githubusercontent.com/2800360/124753663-8c928d80-df29-11eb-9cb1-1ab65d5a093d.png)

I would be fine with solution B. In that case, would you not want to add some additional info to the feature so that it's clear that it's part of another feature, for example

```
 CDS             complement(join(3081..3872,3873..3941))
                 /product="beta-lactamase"
                 [...]
 misc_feature    complement(3873..3941)
                 /label="signal sequence"
                 /product="beta-lactamase"
                 [...]
```

I don't really know whether adding that extra info would be a good idea or not. You know more about the GenBank model than I do, I'll let you decide.

**Comment by gouttegd** [comment]
OK for solution B. There will be a PR soon. :)

As for adding *something* that would allow to "attach" the added feature to the original compound feature: I am not keen on that since all I can foresee is complications no matter what is the *something* we choose to add.

A) We copy one or several qualifier(s) from the original compound feature (as you did in your example with `product`). Which qualifier(s) to copy? What if the original feature has no qualifiers? There is also the risk that a qualifier from the original feature may become misleading when applied to only a part of it (e.g., I think it would be wrong to say that the product of the signal sequence of the *bla* gene is the β-lactamase).

B) We add a `note` that reads like `This is part of feature X`. Where do we get the `X`? From the label of the original feature? Not all features have a label, and conversely there may be several features with the same label (as far as I know there is no rule saying that labels should be unique within a given sequence).

C) We add a `note` that reads like `This is part of the compound feature located in X..Y`. That would be the only always-applicable, non-ambiguous way of doing it. But then this is completely redundant with the information already provided in the feature location, so there's no point…

**Comment by peterjc** [comment]
We had similar issues with trying to map the GFF nested features to the flat list (with only implicit links between gene and CDS etc) approach based on GenBank/EMBL files. We've not come back to revisit that - Brad Chapman did a lot of work in this area, but he and I have not had time or direct need to work on it.

The approach you've outlined sounds reasonable (but I have not looked at the raw representation in SnapGene format), again the co-location and sorting giving context and an implicit link.

**Comment by nicolazilio** [comment]
@gouttegd, I was thinking more about this. At the end of the day, in the case of compound features, I think that besides the coordinates of each part, what people would mostly care about is the name/label of each part. To preserve this information, could one not add a qualifier, for instance called parts, which contains the list of part names (e.g. as a comma-separated list)? I don't know whether this is a good idea or not, I'm just trying to suggest possible solutions to the problem, not to lose information from the original file.

**Comment by gouttegd** [comment]
As far as I know Biopython only uses “standard” qualifiers (those defined in [annex 7.2](http://www.insdc.org/files/feature_table.html#7.2) of the “DDBJ/ENA/GenBank Feature Table Definition”), I am not very keen on introducing a non-standard qualifier…

At the very least, if we go that route this should probably be discussed further with other Biopython developers, and maybe also with other Bio* projects as well (so that there’s a chance of agreeing on a common qualifier instead of having each project coming up with its own hack).

But as it stands, I am in favour of keeping ”Solution B” as described above (and as implemented in the linked PR).

@peterjc Any thoughts on this? Is there an official policy somewhere regarding the use of “standard” vs “non-standard” qualifiers?

**Comment by peterjc** [comment]
We already have examples on non-INSDC qualifiers (ie not used in DDBJ/ENA/GenBank Feature Table Definition), e.g. from IMGT, SwissProt, UniProt

**Comment by gouttegd** [comment]
Ah, that's good to know, thanks. I don't know where I got the impression that we were sticking to the GenBank model.

I think I am still in favour of the "named sub-features are distinct features" approach (aka "Solution B"), but if we're not restricted to INSDC qualifiers then a dedicated qualifier may be a good approach as well (I'd still like that better than using the `note` qualifier for everything).

@nicolazilio Rather than a simple comma-separated list, I'd suggest something like the following:

```
/parts="1:Name of first part;2:Name of second part;7:Name of seventh part"
```

which I think would make it easier to find the name of a part for features made of a large number of parts.

Or maybe

```
/parts="1:Name of first part"
/parts="2:Name of second part"
/parts="7:Name of seventh part"
```

which would be even easier to read.

**Comment by nicolazilio** [comment]
Great! I think "option 1" looks very good, and def better than a simple comma-separated list.

**Comment by gouttegd** [comment]
“Option 1” (`/parts="1:Name of first part;…"`) has been implemented in #3662. :)

**Comment by nicolazilio** [comment]
Thanks a lot!
